File: TimberCruisingApp/Screens/TreeIdentitySheet.swift
// Pre-scan tree identity picker.
//
// Cruiser feedback: "DBH랑 Height를 따로 재면 같은 나무인지 다른
// 나무인지 모르잖아? 재기 전에 미리 tree를 추가하는지, 기존에 있는
// tree에서 정보를 업데이트 하는지 선택하게 해야 할 듯".
//
// This sheet appears between the cruiser tapping a measurement card
// (Diameter / Height) on the Quick Measure home and the actual scan
// cover opening. It offers three choices:
//
//   1. Continue last tree (#N) — one-tap quick path, the most common
//      case when the cruiser is doing DBH then Height back-to-back
//      on the same stem.
//   2. New tree (#N+1) — auto-incremented, also one-tap.
//   3. Pick from existing — opens an inline picker listing every
//      distinct tree number with the readings already attached.
//
// The chosen tree number is passed back via `onPick` and ends up on
// the QuickMeasureEntry that the scan cover writes into history.

import SwiftUI

public struct TreeIdentitySheet: View {

    /// What the sheet is gating: a Diameter scan, a Height scan, or
    /// (later) something else. Shown in the header copy so the
    /// cruiser knows what they're about to start.
    public enum ScanKind {
        case diameter
        case height

        var title: String {
            switch self {
            case .diameter: return "Diameter scan"
            case .height:   return "Height scan"
            }
        }

        var subtitle: String {
            switch self {
            case .diameter: return "Pick which tree this reading is for."
            case .height:   return "Pick which tree this reading is for."
            }
        }
    }

    let scanKind: ScanKind
    let history: QuickMeasureHistory
    let onPick: (Int) -> Void
    let onCancel: () -> Void

    @State private var showAllTrees = false
    @Environment(\.dismiss) private var dismiss

    public init(scanKind: ScanKind,
                history: QuickMeasureHistory,
                onPick: @escaping (Int) -> Void,
                onCancel: @escaping () -> Void) {
        self.scanKind = scanKind
        self.history = history
        self.onPick = onPick
        self.onCancel = onCancel
    }

    public var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: ForestixSpace.md) {
                    header
                    primaryChoices
                    if showAllTrees {
                        allTreesSection
                    }
                }
                .padding(.horizontal, ForestixSpace.md)
                .padding(.top, ForestixSpace.sm)
                .padding(.bottom, ForestixSpace.xl)
            }
            .background(ForestixPalette.canvas.ignoresSafeArea())
            .navigationTitle(scanKind.title)
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") {
                        onCancel()
                        dismiss()
                    }
                }
            }
        }
    }

    // MARK: - Header

    private var header: some View {
        Text(scanKind.subtitle)
            .font(ForestixType.body)
            .foregroundStyle(ForestixPalette.textSecondary)
            .padding(.top, ForestixSpace.xs)
    }

    // MARK: - Primary choices

    /// Two-row stack: continue last tree, or start a new one. These
    /// are the two paths a cruiser hits 95 % of the time. A third
    /// "Pick from existing" row expands a list when tapped.
    private var primaryChoices: some View {
        VStack(spacing: ForestixSpace.sm) {
            if let last = history.lastTreeNumber {
                ChoiceRow(
                    title: "Continue tree #\(last)",
                    subtitle: history.summary(forTreeNumber: last)
                        ?? "Add another reading to this tree",
                    systemImage: "arrow.triangle.2.circlepath",
                    accent: ForestixPalette.primary,
                    accessibilityId: "treeIdentity.continueLast"
                ) {
                    onPick(last)
                    dismiss()
                }
            }
            ChoiceRow(
                title: "New tree #\(history.suggestedNextTreeNumber)",
                subtitle: "Start a fresh tree on this reading",
                systemImage: "plus.circle",
                accent: ForestixPalette.primary,
                accessibilityId: "treeIdentity.newTree"
            ) {
                onPick(history.suggestedNextTreeNumber)
                dismiss()
            }
            if !history.distinctTreeNumbers.isEmpty {
                Button {
                    showAllTrees.toggle()
                } label: {
                    HStack {
                        Text(showAllTrees
                             ? "Hide all trees"
                             : "Pick from existing trees")
                            .font(ForestixType.bodyBold)
                            .foregroundStyle(ForestixPalette.primary)
                        Spacer()
                        Image(systemName: showAllTrees
                              ? "chevron.up"
                              : "chevron.down")
                            .font(.system(size: 13, weight: .semibold))
                            .foregroundStyle(ForestixPalette.primary)
                    }
                    .padding(.horizontal, ForestixSpace.md)
                    .padding(.vertical, ForestixSpace.sm)
                    .frame(maxWidth: .infinity)
                    .background(
                        RoundedRectangle(cornerRadius: ForestixRadius.card,
                                         style: .continuous)
                            .stroke(ForestixPalette.primary.opacity(0.4),
                                    lineWidth: 0.75))
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("treeIdentity.pickExisting")
            }
        }
    }

    // MARK: - All trees expanded list

    private var allTreesSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            sectionHeader("ALL TREES")
            VStack(spacing: 0) {
                ForEach(history.distinctTreeNumbers, id: \.self) { n in
                    Button {
                        onPick(n)
                        dismiss()
                    } label: {
                        HStack(spacing: ForestixSpace.sm) {
                            Text("#\(n)")
                                .font(ForestixType.dataSmall)
                                .foregroundStyle(ForestixPalette.textPrimary)
                                .frame(width: 44, alignment: .leading)
                            Text(history.summary(forTreeNumber: n) ?? "—")
                                .font(ForestixType.caption)
                                .foregroundStyle(ForestixPalette.textSecondary)
                            Spacer()
                            Image(systemName: "chevron.forward")
                                .font(.system(size: 12, weight: .semibold))
                                .foregroundStyle(ForestixPalette.textTertiary)
                        }
                        .padding(.horizontal, ForestixSpace.md)
                        .padding(.vertical, ForestixSpace.sm)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    if n != history.distinctTreeNumbers.last {
                        Rectangle()
                            .fill(ForestixPalette.divider)
                            .frame(height: 0.5)
                            .padding(.leading, ForestixSpace.md + 44 + ForestixSpace.sm)
                    }
                }
            }
            .background(
                RoundedRectangle(cornerRadius: ForestixRadius.card,
                                 style: .continuous)
                    .fill(ForestixPalette.surface))
        }
    }

    private func sectionHeader(_ text: String) -> some View {
        Text(text)
            .font(ForestixType.sectionHead)
            .tracking(1.2)
            .foregroundStyle(ForestixPalette.textTertiary)
            .padding(.bottom, ForestixSpace.xs)
            .padding(.leading, ForestixSpace.xs)
    }
}

// MARK: - Choice row component

private struct ChoiceRow: View {
    let title: String
    let subtitle: String
    let systemImage: String
    let accent: Color
    let accessibilityId: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: ForestixSpace.md) {
                ZStack {
                    RoundedRectangle(cornerRadius: ForestixRadius.control,
                                     style: .continuous)
                        .fill(accent.opacity(0.15))
                        .frame(width: 40, height: 40)
                    Image(systemName: systemImage)
                        .font(.system(size: 17, weight: .medium))
                        .foregroundStyle(accent)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(ForestixType.bodyBold)
                        .foregroundStyle(ForestixPalette.textPrimary)
                    Text(subtitle)
                        .font(ForestixType.caption)
                        .foregroundStyle(ForestixPalette.textSecondary)
                        .lineLimit(2)
                }
                Spacer(minLength: 0)
                Image(systemName: "chevron.forward")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(ForestixPalette.textTertiary)
            }
            .padding(ForestixSpace.md)
            .frame(maxWidth: .infinity)
            .background(
                RoundedRectangle(cornerRadius: ForestixRadius.card,
                                 style: .continuous)
                    .fill(ForestixPalette.surface))
            .overlay(
                RoundedRectangle(cornerRadius: ForestixRadius.card,
                                 style: .continuous)
                    .stroke(ForestixPalette.divider, lineWidth: 0.5))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(accessibilityId)
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/tree/TreeIdentitySheet.kt
// Port of iOS Screens/TreeIdentitySheet.swift.
// Pre-scan tree identity picker.
//
// Cruiser feedback: "DBH랑 Height를 따로 재면 같은 나무인지 다른
// 나무인지 모르잖아? 재기 전에 미리 tree를 추가하는지, 기존에 있는
// tree에서 정보를 업데이트 하는지 선택하게 해야 할 듯".
//
// This sheet appears between the cruiser tapping a measurement card
// (Diameter / Height) on the Quick Measure home and the actual scan
// opening. It offers three choices:
//
//   1. Continue last tree (#N) — one-tap quick path, the most common
//      case when the cruiser is doing DBH then Height back-to-back
//      on the same stem.
//   2. New tree (#N+1) — auto-incremented, also one-tap.
//   3. Pick from existing — opens an inline picker listing every
//      distinct tree number with the readings already attached.
//
// The chosen tree number is passed back via `onPick` and ends up on
// the QuickMeasureEntry that the scan screen writes into history.
//
// Host inside a ModalBottomSheet (see ReconCruiseScreen for the pattern);
// the caller dismisses the sheet from both callbacks.

package com.hcjeong.forestix.ui.screens.tree

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.AddCircleOutline
import androidx.compose.material.icons.filled.Autorenew
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import com.hcjeong.forestix.data.QuickMeasureHistory
import com.hcjeong.forestix.ui.clickableNoRipple
import com.hcjeong.forestix.ui.theme.Forestix
import com.hcjeong.forestix.ui.theme.ForestixRadius
import com.hcjeong.forestix.ui.theme.ForestixSpace

/// What the sheet is gating: a Diameter scan, a Height scan, or
/// (later) something else. Shown in the header copy so the
/// cruiser knows what they're about to start.
enum class ScanKind {
    DIAMETER,
    HEIGHT;

    val title: String
        get() = when (this) {
            DIAMETER -> "Diameter scan"
            HEIGHT -> "Height scan"
        }

    val subtitle: String
        get() = when (this) {
            DIAMETER -> "Pick which tree this reading is for."
            HEIGHT -> "Pick which tree this reading is for."
        }
}

@Composable
fun TreeIdentitySheet(
    scanKind: ScanKind,
    history: QuickMeasureHistory,
    onPick: (Int) -> Unit,
    onCancel: () -> Unit,
) {
    val colors = Forestix.colors
    val type = Forestix.type
    val entries by history.entries.collectAsState()
    var showAllTrees by remember { mutableStateOf(false) }

    /// Most recently used tree number across the log (iOS
    /// `history.lastTreeNumber`; entries are newest-first).
    val lastTreeNumber = entries.firstOrNull { it.treeNumber != null }?.treeNumber
    val distinctTreeNumbers = history.distinctTreeNumbers

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = ForestixSpace.md)
            .padding(top = ForestixSpace.sm, bottom = ForestixSpace.xl),
        verticalArrangement = Arrangement.spacedBy(ForestixSpace.md),
    ) {
        // Title bar (iOS inline nav title + Cancel toolbar button).
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text(scanKind.title, style = type.bodyBold, color = colors.textPrimary)
            Spacer(Modifier.weight(1f))
            TextButton(onClick = onCancel) { Text("Cancel") }
        }

        // MARK: Header
        Text(
            scanKind.subtitle,
            style = type.body,
            color = colors.textSecondary,
        )

        // MARK: Primary choices
        //
        // Two-row stack: continue last tree, or start a new one. These
        // are the two paths a cruiser hits 95 % of the time. A third
        // "Pick from existing" row expands a list when tapped.
        Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.sm)) {
            lastTreeNumber?.let { last ->
                ChoiceRow(
                    title = "Continue tree #$last",
                    subtitle = history.summary(last) ?: "Add another reading to this tree",
                    icon = Icons.Filled.Autorenew,
                ) { onPick(last) }
            }
            ChoiceRow(
                title = "New tree #${history.suggestedNextTreeNumber}",
                subtitle = "Start a fresh tree on this reading",
                icon = Icons.Filled.AddCircleOutline,
            ) { onPick(history.suggestedNextTreeNumber) }
            if (distinctTreeNumbers.isNotEmpty()) {
                Row(
                    Modifier
                        .fillMaxWidth()
                        .clip(ForestixRadius.card)
                        .border(0.75.dp, colors.primary.copy(alpha = 0.4f), ForestixRadius.card)
                        .clickableNoRipple { showAllTrees = !showAllTrees }
                        .padding(horizontal = ForestixSpace.md, vertical = ForestixSpace.sm),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        if (showAllTrees) "Hide all trees" else "Pick from existing trees",
                        style = type.bodyBold,
                        color = colors.primary,
                    )
                    Spacer(Modifier.weight(1f))
                    Icon(
                        if (showAllTrees) Icons.Filled.KeyboardArrowUp
                        else Icons.Filled.KeyboardArrowDown,
                        contentDescription = null,
                        tint = colors.primary,
                        modifier = Modifier.size(18.dp),
                    )
                }
            }
        }

        // MARK: All trees expanded list
        if (showAllTrees) {
            Column {
                Text(
                    "ALL TREES",
                    style = type.sectionHead,
                    color = colors.textTertiary,
                    modifier = Modifier.padding(bottom = ForestixSpace.xs, start = ForestixSpace.xs),
                )
                Column(
                    Modifier
                        .fillMaxWidth()
                        .clip(ForestixRadius.card)
                        .background(colors.surface),
                ) {
                    distinctTreeNumbers.forEachIndexed { index, n ->
                        Row(
                            Modifier
                                .fillMaxWidth()
                                .clickableNoRipple { onPick(n) }
                                .padding(horizontal = ForestixSpace.md, vertical = ForestixSpace.sm),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(ForestixSpace.sm),
                        ) {
                            Text(
                                "#$n",
                                style = type.dataSmall,
                                color = colors.textPrimary,
                                modifier = Modifier.width(44.dp),
                            )
                            Text(
                                history.summary(n) ?: "—",
                                style = type.caption,
                                color = colors.textSecondary,
                            )
                            Spacer(Modifier.weight(1f))
                            Icon(
                                Icons.AutoMirrored.Filled.KeyboardArrowRight,
                                contentDescription = null,
                                tint = colors.textTertiary,
                                modifier = Modifier.size(16.dp),
                            )
                        }
                        if (index != distinctTreeNumbers.lastIndex) {
                            HorizontalDivider(
                                color = colors.divider,
                                thickness = 0.5.dp,
                                modifier = Modifier.padding(
                                    start = ForestixSpace.md + 44.dp + ForestixSpace.sm),
                            )
                        }
                    }
                }
            }
        }
    }
}

// MARK: - Choice row component

@Composable
private fun ChoiceRow(
    title: String,
    subtitle: String,
    icon: ImageVector,
    action: () -> Unit,
) {
    val colors = Forestix.colors
    val type = Forestix.type
    Row(
        Modifier
            .fillMaxWidth()
            .clip(ForestixRadius.card)
            .background(colors.surface)
            .border(0.5.dp, colors.divider, ForestixRadius.card)
            .clickableNoRipple(action)
            .padding(ForestixSpace.md),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(ForestixSpace.md),
    ) {
        Box(
            Modifier
                .size(40.dp)
                .clip(ForestixRadius.control)
                .background(colors.primary.copy(alpha = 0.15f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, contentDescription = null, tint = colors.primary, modifier = Modifier.size(20.dp))
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(title, style = type.bodyBold, color = colors.textPrimary)
            Text(subtitle, style = type.caption, color = colors.textSecondary, maxLines = 2)
        }
        Icon(
            Icons.AutoMirrored.Filled.KeyboardArrowRight,
            contentDescription = null,
            tint = colors.textTertiary,
            modifier = Modifier.size(16.dp),
        )
    }
}
